### PR TITLE
[5] list custom field - max results from xml

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue # .

### Summary of Changes
control max results and max render from fancy-select by xml

### Testing Instructions
Create an SQL field.

Use a query that returns more than 10 values that share the same portion of text for "text," for example:

```sql
SELECT 1 AS value, 'month-January' AS text
UNION ALL
SELECT 2 AS value, 'month-February' AS text
UNION ALL
SELECT 3 AS value, 'month-March' AS text
UNION ALL
SELECT 4 AS value, 'month-April' AS text
UNION ALL
SELECT 5 AS value, 'month-May' AS text
UNION ALL
SELECT 6 AS value, 'month-June' AS text
UNION ALL
SELECT 7 AS value, 'month-July' AS text
UNION ALL
SELECT 8 AS value, 'month-August' AS text
UNION ALL
SELECT 9 AS value, 'month-September' AS text
UNION ALL
SELECT 10 AS value, 'month-October' AS text
UNION ALL
SELECT 11 AS value, 'month-November' AS text
UNION ALL
SELECT 12 AS value, 'month-December' AS text;
```

In the Form Layout set as a view "Enhanced Select".

### Actual result BEFORE applying this Pull Request
When you try to search for "month" as a value in the field, you only display 10 results as per choice.js default 

### Expected result AFTER applying this Pull Request
When you try to search for "month" as a value in the field, the number of values returned is governed by the parameters in the Field -> General
image

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: administrator/language/en-GB/lib_joomla.ini [hunk 1]
```diff
@@ -302,6 +302,8 @@ JLIB_FORM_FIELD_PARAM_INTEGER_STEP_DESC="Each option will be the previous option
 JLIB_FORM_FIELD_PARAM_INTEGER_STEP_LABEL="Step"
 JLIB_FORM_FIELD_PARAM_LAYOUT_BASIC_SELECT="HTML Select"
 JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT="Enhanced Select"
+JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_MAXRENDER="Max render"
```
Comment (brianteeman) [inline administrator/language/en-GB/lib_joomla.ini]: Please follow the style guide

Comment (alikon) [reply administrator/language/en-GB/lib_joomla.ini]: done

File: plugins/fields/sql/params/sql.xml [hunk 2]
```diff
@@ -29,22 +29,63 @@
...
+				label="JLIB_FORM_FIELD_PARAM_LAYOUT_LABEL"
+				class="form-select"
+				default="joomla.form.field.list"
+				>
+				<option value="joomla.form.field.list">JLIB_FORM_FIELD_PARAM_LAYOUT_BASIC_SELECT</option>
+				<option value="joomla.form.field.list-fancy-select">JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT</option>
+			</field>
+
+			<field
+					name="data-max-results"
```
Comment (Quy) [inline plugins/fields/sql/params/sql.xml]: Reduce indentation.

File: plugins/fields/sql/sql.xml [hunk 3]
```diff
@@ -54,16 +54,62 @@
...
 				<field
-					name="form_layout"
+					name="layout"
 					type="list"
 					label="JLIB_FORM_FIELD_PARAM_LAYOUT_LABEL"
 					class="form-select"
-					default="joomla.form.field.list"				      
-					>
+					default="joomla.form.field.list"
+				>
```
Comment (Quy) [inline plugins/fields/sql/sql.xml]:
```suggestion
					>
```

File: plugins/fields/sql/params/sql.xml [hunk 4]
```diff
@@ -29,22 +29,63 @@
...
+					type="list"
```
Comment (Quy) [inline plugins/fields/sql/params/sql.xml:49]: Should be in the following order: name, type, label

File: plugins/fields/sql/params/sql.xml (hunk 2 above)
Comment (Quy) [reply plugins/fields/sql/params/sql.xml]: See line 33.

File: plugins/fields/sql/params/sql.xml (hunk 4 above)
Comment (alikon) [reply plugins/fields/sql/params/sql.xml:49]: done

File: plugins/fields/sql/sql.xml [hunk 5]
```diff
@@ -54,16 +54,62 @@
...
+					<option value="250">J250</option>
+					<option value="300">J300</option>
+					<option value="500">J500</option>
+				</field>
+
+				<field
+					name="data-max-render"
+					type="list"
+					default="-1"
+					label="JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT_MAXRENDER"
```
Comment (Quy) [inline plugins/fields/sql/sql.xml]:
```suggestion
					type="list"
					label="JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT_MAXRESULTS"
					default="10"
					filter="integer"
					validate="options"
					showon="layout:joomla.form.field.list-fancy-select"
					>
					<option value="5">J5</option>
					<option value="10">J10</option>
					<option value="15">J15</option>
					<option value="20">J20</option>
					<option value="25">J25</option>
					<option value="30">J30</option>
					<option value="50">J50</option>
					<option value="100">J100</option>
					<option value="200">J200</option>
					<option value="250">J250</option>
					<option value="300">J300</option>
					<option value="500">J500</option>
				</field>

				<field
					name="data-max-render"
					type="list"
					label="JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT_MAXRENDER"
					default="-1"
```

**Comment by brianteeman** [comment]
why does max-render have JALL but maxresults only have J500

**Comment by alikon** [comment]
if i read correctly choice.js code there is no -1 for for max-results
but i could be wrong

**Comment by brianteeman** [comment]
maybe so but both fields are the same type and layout

**Comment by alikon** [comment]
not sure to get your point
but  500 max search results are far better than 10
or not ?
❓

**Comment by brianteeman** [comment]
I am just querying why we have different settings

**Comment by Razzo1987** [comment]
> I am just querying why we have different settings

The current JS setting (without the options) is:
- Max Result = 10
- Max Render = -1 (ALL)

Max Result does not support -1 (ALL) so it cannot be included in the list because it would not work.
Taking -1 (ALL) out of the Max Render list would limit the number of elements to a maximum of 500, so it would be a loss of functionality.
The only solution I see is to leave the 2 different lists.

**Comment by brianteeman** [comment]
Thank you for a logical answer

**Comment by ceford** [comment]
I have tested this item :white_check_mark: successfully on 760c0295eb010240ad957afc5787aa7aa06a539f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42765">issues.joomla.org/tracker/joomla-cms/42765</a>.</sub>

File: plugins/fields/sql/params/sql.xml [hunk 6]
```diff
@@ -29,22 +29,63 @@
...
-					label="JLIB_FORM_FIELD_PARAM_LAYOUT_LABEL"
-					class="form-select"
-					>
-					<option value="">COM_FIELDS_FIELD_USE_GLOBAL</option>
-					<option value="joomla.form.field.list">JLIB_FORM_FIELD_PARAM_LAYOUT_BASIC_SELECT</option>
-					<option value="joomla.form.field.list-fancy-select">JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT</option>
-				</field>
-			</fieldset>
+			<field
+				name="layout"
```
Comment (Quy) [inline plugins/fields/sql/params/sql.xml:33]: Why renamed?

**Comment by robertolongo** [comment]
I have tested this item :white_check_mark: successfully on 760c0295eb010240ad957afc5787aa7aa06a539f<br /><br />OK, it works.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42765">issues.joomla.org/tracker/joomla-cms/42765</a>.</sub>

**Comment by faustonenci** [comment]
I have tested this item :white_check_mark: successfully on 760c0295eb010240ad957afc5787aa7aa06a539f<br /><br />all right <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42765">issues.joomla.org/tracker/joomla-cms/42765</a>.</sub>

**Comment by crimle** [comment]
I have tested this item :red_circle: unsuccessfully on 760c0295eb010240ad957afc5787aa7aa06a539f<br /><br />Changing firefox to dark mode immediately changes to dark mode in Joomla as well. With or without this patch.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42765">issues.joomla.org/tracker/joomla-cms/42765</a>.</sub>

**Comment by richard67** [comment]
> I have tested this item 🔴 unsuccessfully on [760c029](https://github.com/joomla/joomla-cms/commit/760c0295eb010240ad957afc5787aa7aa06a539f)Changing firefox to dark mode immediately changes to dark mode in Joomla as well. With or without this patch.

@crimle Are you sure you posted your test result for the right pull request (PR)? I don't see how this PR here is related to dark mode. Please check and report back, and either alter your test result back to "not tested" or ask us to do so for you. Thanks in advance.

**Comment by crimle** [comment]
I have not tested this item.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42765">issues.joomla.org/tracker/joomla-cms/42765</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42765">issues.joomla.org/tracker/joomla-cms/42765</a>.</sub>

**Comment by bembelimen** [comment]
I wonder, why did you rename the field name? How do you take care about the "old" values, isn't this a B/C break then?

**Comment by alikon** [comment]
iirc i didn't want to touch com_fields
to limit the impact surface of the pr

**Comment by Razzo1987** [comment]
> I wonder, why did you rename the field name? How do you take care about the "old" values, isn't this a B/C break then?

Leaving `form_layout` and binding the 2 options to that tag does not work.  
`data-*` options are loaded only with `layout`, this xml tag is also documented [here](https://github.com/joomla/joomla-cms/pull/32591), when fancy was introduced.

To make it work with `form_layout` com_fields has to be rewritten, which was not intended to be the goal of this PR

**Comment by Fedik** [comment]
Not ready<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42765">issues.joomla.org/tracker/joomla-cms/42765</a>.</sub>

**Comment by Fedik** [comment]
The PR is not finished.
It can go in 5.x, if done properly, there no b/c break.

Current XML modifications should be reverted. As we talked with @Razzo1987 
And some more changes still need to be done.

**Comment by Fedik** [comment]
I would add, that the changes should be done for `list` field first, and then copyed to `sql` field.

**Comment by Razzo1987** [comment]
Hi @Fedik, the problem stems from the fact that to fix the bug and continue using `form_layout` the change on com_fields should be made. If I understand correctly from @alikon he does not feel confident in touching that part. If you can give him support I am also confident that using `form_layout` avoids b/c breaks.

**Comment by alikon** [comment]
i would like to make this happen for 5.x not 6.x, let's join together the effort then, the pr it's already open, better,  
let's schedule a meet via mattermost on the weekend  ? if you can

**Comment by alikon** [comment]
thank you @Fedik

File: plugins/fields/sql/sql.xml [hunk 7]
```diff
@@ -54,16 +54,62 @@
...
+				<field
+					name="data-max-results"
+					type="list"
+					label="JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT_MAXRESULTS"
+					default="10"
+					filter="integer"
+					validate="options"
+					showon="layout:joomla.form.field.list-fancy-select"
+					>
+					<option value="5">J5</option>
```
Comment (Quy) [inline plugins/fields/sql/sql.xml:76]: Should these options match the other xml files?

Comment (Fedik) [reply plugins/fields/sql/sql.xml:76]: Which other? I do not see a reason to add a loot of them, here

Comment (Quy) [reply plugins/fields/sql/sql.xml:76]: My suggestion is to reduce options here to match in your PR.

Comment (alikon) [reply plugins/fields/sql/sql.xml:76]: done

File: plugins/fields/sql/sql.xml [hunk 8]
```diff
@@ -54,16 +54,48 @@
...
-					default="joomla.form.field.list"				      
+					default="joomla.form.field.list"
 					>
 					<option value="joomla.form.field.list">JLIB_FORM_FIELD_PARAM_LAYOUT_BASIC_SELECT</option>
 					<option value="joomla.form.field.list-fancy-select">JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT</option>
 				</field>
-				
+
+				<field
+					name="data-max-results"
```
Comment (Fedik) [inline plugins/fields/sql/sql.xml]: please change the name to `snake_case`, as it done in the filed params form

File: plugins/fields/list/params/list.xml [hunk 9]
```diff
@@ -60,6 +60,37 @@
...
+				</field>
+				<field
+					name="data_max_render"
+					type="list"
+					label="JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT_MAXRENDER"
+					default="-1"
+					filter="integer"
+					validate="options"
+					showon="form_layout:joomla.form.field.list-fancy-select"
+				>
```
Comment (Quy) [inline plugins/fields/list/params/list.xml]:
```suggestion
					>
					<option value="5">J5</option>
					<option value="10">J10</option>
					<option value="25">J25</option>
					<option value="50">J50</option>
					<option value="100">J100</option>
				</field>
				<field
					name="data_max_render"
					type="list"
					label="JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT_MAXRENDER"
					default="-1"
					filter="integer"
					validate="options"
					showon="form_layout:joomla.form.field.list-fancy-select"
					>
```

File: plugins/fields/sql/params/sql.xml [hunk 10]
```diff
@@ -44,6 +44,37 @@
...
+				</field>
+				<field
+					name="data_max_render"
+					type="list"
+					label="JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT_MAXRENDER"
+					default="-1"
+					filter="integer"
+					validate="options"
+					showon="form_layout:joomla.form.field.list-fancy-select"
+				>
```
Comment (Quy) [inline plugins/fields/sql/params/sql.xml]:
```suggestion
					>
					<option value="5">J5</option>
					<option value="10">J10</option>
					<option value="25">J25</option>
					<option value="50">J50</option>
					<option value="100">J100</option>
				</field>
				<field
					name="data_max_render"
					type="list"
					label="JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT_MAXRENDER"
					default="-1"
					filter="integer"
					validate="options"
					showon="form_layout:joomla.form.field.list-fancy-select"
					>
```

File: administrator/components/com_fields/src/Plugin/FieldsListPlugin.php [hunk 11]
```diff
@@ -53,6 +53,18 @@ public function onCustomFieldsPrepareDom($field, \DOMElement $parent, Form $form
             $element->setAttribute('value', $value);
         }
 
+        // Set data attributes
+        $dataMaxResult = $field->params->get('data_max_results', 10);
+        $dataMaxRender = $field->params->get('data_max_render', -1);
+
+        if ($dataMaxResult) {
```
Comment (Quy) [inline administrator/components/com_fields/src/Plugin/FieldsListPlugin.php]: Is this check necessary rather than just set the attribute?

Comment (alikon) [reply administrator/components/com_fields/src/Plugin/FieldsListPlugin.php]: no

**Comment by Fedik** [comment]
> However, I propose to include j500 in the options.

It will have a huge performance and usability issue, I think anything more than 20-30 is already to much.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on cdc2480aee53b81ada80e00e25280450d5c024bb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42765">issues.joomla.org/tracker/joomla-cms/42765</a>.</sub>

**Comment by Razzo1987** [comment]
> > However, I propose to include j500 in the options.
> 
> It will have a huge performance and usability issue, I think anything more than 20-30 is already to much.

I'll try to do a video, but I really don't think it changes much in terms of performance.
I have a SQL field that extracts 1500+ records and the latency is related to the query not the limits imposed by the fancy-select

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on cdc2480aee53b81ada80e00e25280450d5c024bb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42765">issues.joomla.org/tracker/joomla-cms/42765</a>.</sub>

**Comment by Razzo1987** [comment]
### Performance test

Create a SQL field with multiple values, for example:
```sql
SELECT concat(`TABLE_SCHEMA`, '-', `TABLE_NAME`, '-', `COLUMN_NAME`) AS value,
               concat(`TABLE_SCHEMA`, '.', `TABLE_NAME`, '.', `COLUMN_NAME`) AS text
FROM `information_schema`.`COLUMNS`
```

**Test 1**: Form Layout = HTML Select


https://github.com/joomla/joomla-cms/assets/906604/9b034ad6-7374-41f9-a7c6-5316167e4803


**Test 2**: Form Layout = Enhanced Select, Max Result = 10, Max Render = All


https://github.com/joomla/joomla-cms/assets/906604/d824a154-0c62-40c0-954e-05bdf6c5c3f9


**Test 3**: Form Layout = Enhanced Select, Max Result = 500, Max Render = All


https://github.com/joomla/joomla-cms/assets/906604/a4323769-b42d-4ad2-97d2-d3c1e9fed484


Personally, I do not observe significant performance changes in user experience between 10 and 500

**Comment by Razzo1987** [comment]
@alikon I think is missing the new option on `plugins/fields/list/params/list.xml`.
I do the PR on your branch ;)
Before merge check if the name must be "form_layout" or "layout" (in my PR and in this one)

**Comment by Fedik** [comment]
Sorry @Razzo1987 , but 6 second for the form rendering I would call it disaster :wink: 
And imagine User add multiple of this fields?

**Comment by Razzo1987** [comment]
> Sorry @Razzo1987 , but 6 second for the form rendering I would call it disaster 😉 And imagine User add multiple of this fields?
> 
> **Note**: changes in this PR does not affect SQL limit, only the rendering limit for choices.js

Correct, but the 6 seconds is because the form loads 1500+ items in the list.
In fact even with HTML Select the time is that. the fact that I set the search limit to 500 did not affect with other lag

**Comment by Fedik** [comment]
> Correct, but the 6 seconds is because the form loads 1500+ items in the list.

And that what you should never do, ever :wink: 
You may have fast PC, but someone have older, and it may take a minutes just to render.

> the fact that I set the search limit to 500 did not affect with other lag

Yeah, well, but I still do not like it, it just unusable, and fancy-layout allows to search already :wink:

File: plugins/fields/sql/sql.xml [hunk 12]
```diff
@@ -54,16 +54,48 @@
 				</field>
 
 				<field
-					name="form_layout"
+					name="layout"
```
Comment (Fedik) [inline plugins/fields/sql/sql.xml]: Please use old name

File: plugins/fields/sql/sql.xml [hunk 13]
```diff
@@ -54,16 +54,48 @@
...
-				
+
+				<field
+					name="data_max_results"
+					type="list"
+					label="JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT_MAXRESULTS"
+					default="10"
+					filter="integer"
+					validate="options"
+					showon="layout:joomla.form.field.list-fancy-select"
```
Comment (Fedik) [inline plugins/fields/sql/sql.xml]: showon should use `form_layout:...` instead of `layout:...`

File: plugins/fields/sql/sql.xml [hunk 14]
```diff
@@ -54,16 +54,48 @@
...
+				</field>
+
+				<field
+					name="data_max_render"
+					type="list"
+					label="JLIB_FORM_FIELD_PARAM_LAYOUT_FANCY_SELECT_MAXRENDER"
+					default="-1"
+					filter="integer"
+					validate="options"
+					showon="layout:joomla.form.field.list-fancy-select"
```
Comment (Fedik) [inline plugins/fields/sql/sql.xml]: Same here

File: plugins/fields/sql/sql.xml (hunk 12 above)
Comment (alikon) [reply plugins/fields/sql/sql.xml]: done

File: plugins/fields/sql/sql.xml (hunk 13 above)
Comment (alikon) [reply plugins/fields/sql/sql.xml]: done

File: plugins/fields/sql/sql.xml (hunk 14 above)
Comment (alikon) [reply plugins/fields/sql/sql.xml]: done

**Comment by Razzo1987** [comment]
I have tested this item :white_check_mark: successfully on 058087f2573d776d3849d3ee3ddd742fd6ac4bc7<br /><br />Successfully tested with both List and SQL.

Tested with the example of the 12 months of the year and used the combinations:

* Max Results: 10, Max Render: ALL: He behaves as he always has
* Max Results: 100, Max Render: ALL: Show the 12 months by searching "month" (there are less than 100)
* Max Results: 5, Max Render: 5: Show only 5 months, also by searching "month"
* Max Results: 100, Max Render: 5: Show only 5 months, by search it show all the 12 months

Another test: if I set the fancy layout before applying the patch now it is retained.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42765">issues.joomla.org/tracker/joomla-cms/42765</a>.</sub>

**Comment by robertolongo** [comment]
I made the following tests (using the SQL query with months):

- Max Results: 10 - Max Render: ALL - Search string: empty --> I see the list complete
- Max Results: 10 - Max Render: 5 - Search string: empty --> I see the first 5 elements in the list
- Max Results: 10 - Max Render: ALL - Search string: "month"--> I see the first 10 elements in the list
- Max Results: 25 - Max Render: ALL - Search string: "month" --> I see the list complete


I changed the query to obtain a resultset with 26 elements:

- Max Results: 25 - Max Render: ALL - Search string: "month" -->I see the first 25 elements in the list
- Max Results: 25 - Max Render: 10- Search string: empty -->I see the the first 10 elements in the list
- Max Results: 100 - Max Render: ALL - Search string: "month" -->I see the list complete

The patch worked as expected.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42765">issues.joomla.org/tracker/joomla-cms/42765</a>.</sub>

**Comment by richard67** [comment]
I've restored @Razzo1987 's test result in the issue tracker since the commit which invalidated the test counter was just a clean branch update.

@Quy If you update the branch of a PR which has human test results and the branch update is clean without any merge conflicts to be solved, could you then restore the test results of the users by using the "Alter test" button in the issue tracker? Thanks in advance.

**Comment by Razzo1987** [comment]
> Ah I just see it was ready to commit already.

Yes, it had remained in RTC even after the changes 3 days ago, I don't know why.
I wrote to Roberto today if he could redo the tests that have done during PBF ;)

**Comment by richard67** [comment]
Well we have now 2 tests again which are valid, so RTC is again valid.

**Comment by alikon** [comment]
as there is no interest

**Comment by alikon** [comment]
redone and rebased on 5.3  pr #44794

**Pull request closed without merging**